File: javascript/pig-latin/pig-latin.js
const PigLatin = function() {};

PigLatin.translate = function(input) {
  return input.split(' ').map(PigLatin.pigWord).join(' ');
};

PigLatin.pigWord = function(input) {
  const vowelSounds = ['a', 'e', 'i', 'o', 'u'];
  const consonantSets = ['ch', 'th', 'thr', 'sch', 'qu', 'squ'];
  let startsVowel = false;
  let startLetters = '';
  vowelSounds.forEach((i) => {
    if (input.startsWith(i)) {
      startsVowel = true;
      startLetters = i;
    }
  });
  consonantSets.forEach((i) => {
    if (input.startsWith(i)) {
      startLetters = i;
    }
  });

  if (startsVowel) {
    return `${input}ay`;
  }
  if (startLetters.length == 0) { startLetters = input[0]; }
  return `${input.slice(startLetters.length)}${startLetters}ay`;
};

module.exports = PigLatin;

File: python/pig-latin/pig_latin.py
VOWEL_SOUNDS   = ['a', 'e', 'i', 'o', 'u']
CONSONANT_SETS = ['ch', 'th', 'thr', 'sch', 'qu', 'squ']
NON_STARTERS   = ['yt', 'xr']

def translate(words):
    output = []

    for word in words.split(' '):
        start_letters = 1

        c = False
        for vs in VOWEL_SOUNDS:
            if word.startswith(vs):
                output.append("%say" % (word))
                c = True

        if c: continue

        for cs in CONSONANT_SETS:
            if word.startswith(cs):
                start_letters = len(cs)

        for ns in NON_STARTERS:
            if word.startswith(ns):
                start_letters = 0

        output.append("%s%say" % (word[start_letters:], word[:start_letters]))

    return ' '.join(output)